File: Parte1/Version2/MatMul.java
/*
Multiplicação de Matrizes Quadradas (N x N) — Precisão Simples (FP32)

INSTRUÇÕES DE COMPILAÇÃO E EXECUÇÃO
----------------------------------
1. Certifique-se de ter o Java JDK 11 ou superior:
   $ java -version
   $ javac -version

2. Salve este arquivo como:
   MatMul.java

3. Compile o código:
   $ javac MatMul.java

4. Execute o programa:
   $ java MatMul

OBSERVAÇÕES IMPORTANTES
----------------------
- O algoritmo possui complexidade O(N³).
- O código utiliza precisão simples (FP32), usando o tipo float.
- Para N = 1024, o tempo de execução pode ser de vários minutos,
  dependendo do processador.
- O código executa o kernel principal 4 vezes:
    * a primeira execução é descartada (warm-up)
    * a média é calculada sobre as 3 execuções restantes
- A medição de tempo é feita com System.nanoTime().
*/

import java.util.Random;

public class MatMul {

    // Tamanho da matriz (N x N)
    static final int N = 4096;

    // Número total de execuções
    static final int NUM_EXECUCOES = 2;

    public static void main(String[] args) {

        Random rand = new Random();

        // Matrizes em precisão simples (FP32)
        float[][] A = new float[N][N];
        float[][] B = new float[N][N];

        // Inicialização das matrizes
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                A[i][j] = rand.nextFloat();
                B[i][j] = rand.nextFloat();
            }
        }

        float[] tempos = new float[NUM_EXECUCOES - 1];
        int idxTempo = 0;

        for (int execucao = 0; execucao < NUM_EXECUCOES; execucao++) {

            // Matriz resultado
            float[][] C = new float[N][N];

            long inicio = System.nanoTime();

            for (int i = 0; i < N; i++) {
                for (int j = 0; j < N; j++) {
                    for (int k = 0; k < N; k++) {
                        C[i][j] += A[i][k] * B[k][j];
                    }
                }
            }

            long fim = System.nanoTime();
            float tempoSegundos = (fim - inicio) / 1_000_000_000.0f;

            if (execucao == 0) {
                System.out.printf(
                    "Execução %d: %.6f s (descartada – warm-up)%n",
                    execucao + 1, tempoSegundos
                );
            } else {
                tempos[idxTempo++] = tempoSegundos;
                System.out.printf(
                    "Execução %d: %.6f s%n",
                    execucao + 1, tempoSegundos
                );
            }
        }

        // Cálculo da média
        float soma = 0.0f;
        for (float t : tempos) {
            soma += t;
        }

        float media = soma / tempos.length;

        System.out.printf("%nTempo médio: %.6f s%n", media);
    }
}

File: Parte1/Version1/matmul.py
"""
Multiplicação de Matrizes Quadradas (N x N)

INSTRUÇÕES DE EXECUÇÃO
---------------------
1. Certifique-se de ter Python 3.10 ou superior:
   $ python3 --version

2. Salve este arquivo como, por exemplo:
   matmul.py

3. Execute no terminal:
   $ python3 matmul.py

OBSERVAÇÕES IMPORTANTES
----------------------
- O algoritmo possui complexidade O(N³).
- Para N = 2028, o tempo de execução pode ser de vários minutos.
- O código executa o kernel principal 6 vezes:
    * a primeira execução é descartada (warm-up)
    * a média é calculada sobre as 5 execuções restantes
"""

import random
import time

# Tamanho da matriz (N x N)
n = 2028

# Inicialização das matrizes A e B
A = [[random.random() for _ in range(n)] for _ in range(n)]
B = [[random.random() for _ in range(n)] for _ in range(n)]

# Número total de execuções
num_execucoes = 2
tempos = []

for execucao in range(num_execucoes):
    # Reinicializa a matriz resultado a cada execução
    C = [[0.0 for _ in range(n)] for _ in range(n)]

    # Início da medição de tempo (alta resolução)
    inicio = time.perf_counter()

    # Loop principal da multiplicação de matrizes

    for i in range(n):
        for j in range(n):
            for k in range(n):
                C[i][j] += A[i][k] * B[k][j]

    # Fim da medição de tempo
    fim = time.perf_counter()
    tempo = fim - inicio

    if execucao == 0:
        # Primeira execução descartada (warm-up)
        print(f"Execução {execucao + 1}: {tempo:.6f} s (descartada – warm-up)")
    else:
        tempos.append(tempo)
        print(f"Execução {execucao + 1}: {tempo:.6f} s")

# Cálculo da média das execuções válidas
media = sum(tempos) / len(tempos)

print(f"\nTempo médio: {media:.6f} s")
